doc/develop/test/pytest.rst

```diff
@@ -0,0 +1,84 @@
+``PYTHONPATH`` with path to this plugin, and then during pytest call, it appends the command with
+``-p twister_harness.plugin`` argument. If one prefers to use the installed version of the plugin,
+they must add ``--allow-installed-plugin`` flag to twister’s call.
+
+Pytest-based test suites are discovered the same way as other twister tests, i.e., by a presence
+of testcase/sample.yaml. Inside, a keyword ``harness`` tells twister how to handle a given test.
+In the case of ``harness: pytest``, most of twister workflow (test suites discovery,
+parallelization, building and reporting) remains the same as for other harnesses. The change
+happens during the execution step. The below picture presents a simplified overview of the
+integration.
+
+.. figure:: twister_and_pytest.svg
+   :figclass: align-center
+
+
+If ``harness: pytest`` is used, twister delegates the test execution to pytest, by calling it as
+a subprocess. Required parameters (such as build directory, device to be used, etc.) are passed
+through a CLI command. When pytest is done, twister looks for a pytest report (results.xml) and
+sets the test result accordingly.
+
+How to create a pytest test
+***************************
+
+An example of a pytest test is given at :zephyr_file:`samples/subsys/testsuite/pytest/shell/pytest/test_shell.py`.
+Twister calls pytest for each configuration from the .yaml file which uses ``harness: pytest``.
+By default, it points to ``pytest`` directory, located next to a directory with binary sources.
+A keyword ``pytest_root`` placed under ``harness_config`` section can be used to point to another
+location.
+
+Pytest scans the given folder looking for tests, following its default
+`discovery rules <https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#conventions-for-python-test-discovery>`_
+One can also pass some extra arguments to the pytest from yaml file using ``pytest_args`` keyword
+under ``harness_config``, e.g.: ``pytest_args: [‘-k=test_method’, ‘--log-level=DEBUG’]``.
+
+Two imports are important to include in .py sources:
+
+.. code-block:: python
+
+   import pytest  # noqa # pylint: disable=unused-import
+   from pytest_twister_harness.device.device_abstract import DeviceAbstract
+
+The first enables pytest-twister-harness plugin indirectly, as it is added with pytest.
+It also gives access to ``dut`` fixture. The second is important for type checking and enabling
+IDE hints for duts. The ``dut`` fixture is the core of pytest harness plugin. When used as an
+argument of a test function it gives access to a DeviceAbstract type object. The fixture yields a
+device prepared according to the requested type (native posix, qemu, hardware, etc.). All types of
+devices share the same API. This allows for writing tests which are device-type-agnostic.
+
+
+Limitations
+***********
+
+* Twister options like ``--pre-script``, ``--post-flash-script`` and ``--post-script`` are not handled by
```

Comment:
why not? is this planned? can you please put this in the issue mentioned above?

Reply:
btw, this excludes majority of boards and devices from using pytest, so this is a blocker issue for supporting devices.

Reply:
Can we add this to the "tracking issue"? We looked at this and adding this seems rather straight forward and can be done in a next iteration.

Reply:
why can't this be supported by passing the pre/post scripts as command line options to pytest like everything else and call them from pytest? It should be easy enough to do, you are just calling a script...